lib/src/main/java/com/otaliastudios/transcoder/internal/thumbnails/DefaultThumbnailsEngine.kt

```diff
@@ -250,10 +252,36 @@ class DefaultThumbnailsEngine(
             positions.mapNotNull { (positionUs, request) ->
                 val localizedUs = timer.localize(TrackType.VIDEO, index, positionUs)
                 localizedUs?.let { Stub(request, positionUs, localizedUs) }
-            }.toMutableList().sortedBy { it.positionUs }
+//            }.toMutableList().sortedBy { it.positionUs }
+            }.toMutableList().reorder(dataSources[TrackType.VIDEO][0].keyFrameTimestampsUs)
         )
     }
 
+    private fun  List<Stub>.reorder(keyFrameTimestampsUs: ArrayList<Long>): Collection<Stub> {
+        if(keyFrameTimestampsUs.isEmpty())
+            return this
+        bucketListMap.clear()
+        finalList.clear()
+
+        forEach {
+            val keyFrame = findPrevKeyFrame(it.positionUs, keyFrameTimestampsUs)
+            val list = bucketListMap.getOrPut(keyFrame) { ArrayList<Stub>() }
+            list.add(it)
+        }
+        bucketListMap.forEach {
+            finalList.addAll(it.value.sortedBy { it.positionUs })
+        }
+        return finalList
+    }
+
+    private fun findPrevKeyFrame(positionUs: Long, keyFrameTimestampsUs: ArrayList<Long>): Long {
+        for(i in 0 until keyFrameTimestampsUs.size) {
```

Comment:
we already have a search method.
We should use that.

Reply:
Also, we are relying on search to request and add more keyframes if required.
which means We should not use keyFrameTimestampUs list.

And let's use proper variable names

Reply:
A list cannot have a name like keyFrameTimestampsUs